src/librustc_mir/transform/lower_128bit.rs

```diff
@@ -0,0 +1,160 @@
+    }
+}
+
+impl Lower128Bit {
+    fn lower_128bit_ops<'a, 'tcx>(&self, tcx: TyCtxt<'a, 'tcx, 'tcx>, mir: &mut Mir<'tcx>) {
+        let mut new_blocks = Vec::new();
+        let cur_len = mir.basic_blocks().len();
+
+        let (basic_blocks, local_decls) = mir.basic_blocks_and_local_decls_mut();
+        for block in basic_blocks.iter_mut() {
+            for i in (0..block.statements.len()).rev() {
+                let call_did =
+                    if let Some(call_did) = lower_to(&block.statements[i], local_decls, tcx) {
+                        call_did
+                    } else {
+                        continue;
+                    };
+
+                let after_call = BasicBlockData {
+                    statements: block.statements.drain((i+1)..).collect(),
+                    is_cleanup: block.is_cleanup,
+                    terminator: block.terminator.take(),
+                };
+
+                let bin_statement = block.statements.pop().unwrap();
+                let (source_info, lvalue, lhs, rhs) = match bin_statement {
+                    Statement {
+                        source_info,
+                        kind: StatementKind::Assign(
+                            lvalue,
+                            Rvalue::BinaryOp(_, lhs, rhs))
+                    } => (source_info, lvalue, lhs, rhs),
+                    Statement {
+                        source_info,
+                        kind: StatementKind::Assign(
+                            lvalue,
+                            Rvalue::CheckedBinaryOp(_, lhs, rhs))
+                    } => (source_info, lvalue, lhs, rhs),
+                    _ => bug!("Statement doesn't match pattern any more?"),
+                };
+
+                let bb = BasicBlock::new(cur_len + new_blocks.len());
+                new_blocks.push(after_call);
+
+                block.terminator =
+                    Some(Terminator {
+                        source_info,
+                        kind: TerminatorKind::Call {
+                            func: Operand::function_handle(tcx, call_did,
+                                Slice::empty(), source_info.span),
+                            args: vec![lhs, rhs],
+                            destination: Some((lvalue, bb)),
+                            cleanup: None,
+                        },
+                    });
+            }
+        }
+
+        basic_blocks.extend(new_blocks);
+    }
+}
+
+fn lower_to<'a, 'tcx, D>(statement: &Statement<'tcx>, local_decls: &D, tcx: TyCtxt<'a, 'tcx, 'tcx>)
+    -> Option<DefId>
+    where D: HasLocalDecls<'tcx>
+{
+    match statement.kind {
+        StatementKind::Assign(_, Rvalue::BinaryOp(bin_op, ref lhs, _)) => {
+            let ty = lhs.ty(local_decls, tcx);
+            if let Some(is_signed) = sign_of_128bit(&ty) {
+                if let Some(item) = item_for_op(bin_op, is_signed) {
+                    return Some(tcx.require_lang_item(item))
```

Comment:
Since lang-items aren’t yet type-checked, could you add an assertion of some sort that checks whether the function really has the expected type?

https://github.com/rust-lang/rust/issues/9307 is a long standing issue regarding that, and I would rather have an assertion in an easy to track location rather than right before trans.

Reply:
Ok, the next update generates an error like
```rust
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `[i128, i128, (i128, u32)]`,
 right: `[i128, i128, (i128, bool)]`: lang item _ZN23lower_128bit_debug_test9i128_addo', src\librustc_mir\transform\lower_128bit.rs:115:4
```